File: mavlink/simple_flight_controller_c/src/pid/pid.c
/**
 * @file    pid.c
 * @author  Joel Wachsler
 * @author  Daniel Aros Banda 
 * @date    2018-05-13
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include "pid.h"
#include "../util/common.h"
#include "../mavlink/mavlink_com.h"
#include "../pid/calculate_pids.h"

double calculate_pid(double input, pid* curr_pid) {
  double pid_error_temp = input - curr_pid->setpoint;

  curr_pid->i_mem += curr_pid->i_gain * pid_error_temp;

  if(curr_pid->i_mem > curr_pid->max) curr_pid->i_mem = curr_pid->max;
  else if(curr_pid->i_mem < curr_pid->max * -1) curr_pid->i_mem = curr_pid->max * -1;

  double pid_output = curr_pid->p_gain * pid_error_temp + curr_pid->i_mem + curr_pid->d_gain * (pid_error_temp - curr_pid->last_d_error);

  if(pid_output > curr_pid->max) pid_output = curr_pid->max;
  else if(pid_output < curr_pid->max * -1) pid_output = curr_pid->max * -1;

  curr_pid->last_d_error = pid_error_temp;

  return pid_output;
}

File: mavlink/simple_flight_controller_rust/src/calculate_pids/pid.rs
/// Represents an instance of a PID.
pub struct PID {
    p_gain: f64,
    i_gain: f64,
    d_gain: f64,
    max: f64,
    pub setpoint: f64,
    i_mem: f64,
    last_d_error: f64,
}

impl PID {
    pub fn new(p_gain: f64, i_gain: f64, d_gain: f64, max: f64) -> PID {
        PID {
            p_gain,
            i_gain,
            d_gain,
            max,
            setpoint: 0.0,
            i_mem: 0.0,
            last_d_error: 0.0,
        }
    }

    pub fn calculate_pid(&mut self, input: f64) -> f64 {
        let pid_error_temp = input - self.setpoint;

        self.i_mem += self.i_gain * pid_error_temp;

        if self.i_mem > self.max {
            self.i_mem = self.max;
        } else if self.i_mem < self.max * -1.0 {
            self.i_mem = self.max * -1.0;
        }

        let mut pid_output = self.p_gain * pid_error_temp + self.i_mem + self.d_gain * (pid_error_temp - self.last_d_error);

        if pid_output > self.max {
            pid_output = self.max;
        } else if pid_output < self.max * -1.0 {
            pid_output = self.max * -1.0;
        }

        self.last_d_error = pid_error_temp;

        pid_output
    }
}
